doc/reference/trd106-completion-codes.md

```diff
@@ -0,0 +1,86 @@
+Application Completion Codes
+========================================
+
+**TRD:** 106 <br/>
+**Working Group:** Kernel<br/>
+**Type:** Documentary<br/>
+**Status:** Draft<br/>
+**Author:** Alyssa Haroldsen<br/>
+**Draft-Created:** December 6, 2021<br/>
+**Draft-Modified:** January 25, 2022<br/>
+**Draft-Version:** 1<br/>
+**Draft-Discuss:** tock-dev@googlegroups.com</br>
+
+Abstract
+-------------------------------
+This advisory document describes the expected behavior of application completion
+codes when terminating via the `exit` syscall, as described in
+[TRD 104][exit-syscall].
+
+1 Introduction
+===============================
+When an application exits via the [`exit` syscall][exit-syscall], it can specify
+a **completion code**, an unsigned 32-bit number which indicates status. This
+information can be stored in the kernel and used in management or policy
+decisions.
+
+This number is called an "exit status", "exit code", or "result code" on other
+platforms.
+
+2 Design Considerations
+===============================
+When possible, Tock applications should follow existing conventions and
+terminology from other major platforms. This assists in helping the project be
+more understandable to newcomers by following the principle of least
+astonishment.
+
+This advisory document provides guidance for the ecosystem of Tock applications
+using the `exit` syscall, and does not define the behavior of the syscall
+itself.
+
+3 Design
+===============================
+A completion code of `0` passed to the `exit` syscall MUST indicate normal app
+termination. A non-zero completion code SHOULD be used to indicate abnormal
+termination. This distinction is useful so that a Tock kernel can handle
+success/failure cases differently, e.g. by printing error messages,
+and so that kernel extensions (such as process exit handlers defined by a board)
+or external tools (such as a tool designed to parse the output from a kernel
+with *trace\_syscalls* enabled) can match on these two cases.
+This behavior also matches the convention for Unix exit codes, such that it
+likely matches the expectations for users coming from that domain.
+
+A completion code between `1` and `1024` inclusive SHOULD be the
+same value as one of the error codes specified in [TRD 104][error-codes].
+This requirement is a SHOULD rather than a MUST because it is useful in the
+common case (it allows software to infer something about the cause of an
+error that led to an exit, and possibly print a useful message) but also
+allows a process to do something else if needed (e.g. for compatibility
+with some other standard of exit codes).
+
+Accordingly, the core kernel MUST NOT assume any semantic meaning for completion
+codes or take actions based on their values besides printing error messages.
```

Comment:
```suggestion
Accordingly, the core kernel MUST NOT assume any semantic meaning for completion
codes or take actions based on their values besides printing error messages, as long as
there is no explicit contract between the kernel and application regarding completion
code interpretation.
```

I like that this does imply that the kernel will not take arbitrary action based solely on the completion codes, but depending on how this is read, it does seem to severely hinder the usefulness of return codes if an app _wants_ do define an explicit contract with the kernel. It might be more clear that this is still an option if we add this phrase explicitly.

Reply:
I like this, though wouldn't "core kernel" mean that capsules still have the freedom to assume semantic meaning for completion codes?

Reply:
Yes, I think that interpretation is reasonable. Maybe it's worth adding an emphasis to core, as I didn't really realize capsules were deliberately excluded from this statement while reading.

I still think that, given the scheduler and process fault handler be considered part of the core kernel, _allowing_ an explicit contract between an app and the kernel may make sense.

Reply:
@hudson-ayers, @phil-levis, what are your thoughts on this? I'm fine with the suggested wording here.

Reply:
I am happy with Leon's suggested wording as well

Reply:
Sorry, missed this chain. What's an explicit contract? How would such a contract emerge?

Reply:
@lschuermann @hudson-ayers @kupiakos Just wanted to follow up on this. Would you like me to suggest some text to describe what we talked about today, or Leon, I think you said you'd do that?

Reply:
@phil-levis Sure, didn't get to it until now.

Based on the feedback I suppose it's good to establish my primary motivations behind this change, if only for future record:
- the current wording might be to. restrictive. While downstream users don't technically have to abide to these TRDs, they will the direction of the upstream kernel takes. If we rule out that the kernel can take actions based on app completion codes, we might design internal interfaces that reflect that decision, making it harder for downstream users to implement something like this.
- we _MUST_ ensure that the kernel does not take action on app completion codes as long as there is no guarantee that the kernel knows how to interpret a specific app's completion codes. It is unacceptable for the core kernel to misidentify an app or otherwise misinterpret its completion codes and take action based on this information.

Thus, a more elaborate text might say

```suggestion
Accordingly, the core kernel MUST NOT assume any semantic meaning for completion
codes. Without a revision of this TRD, the upstream core kernel may not take actions
based on their values besides printing error messages. Future versions of the kernel
may implement mechanisms to take actions based on an app's completion code,
provided that
- the app provides a specification for its completion code space, specifically concerning
  completion codes encoding success and failure cases. It may further specify which types
  of actions the core kernel may enact based on its completion codes.
- the kernel adheres to the app's completion code specification and only takes actions
  allowed by the specification based on the completion code returned. This implies
  that an application with a specification that does not include any instructions about
  actions taken in response to a return code will not be different from an app without such an
  application. The kernel may not implement every action described in this specification.
- the kernel can reliably identify the app and associate it with its completion code specification.
```

This is very elaborate on what a future design could and should look like, but it does not yet fully describe how such a specification must look like and what qualifies as a completion-code specification. It does not necessarily have any benefits for us in the upstream Tock codebase right now, other than being able to design internal interfaces to incorporate such future behavior. Still, it provides downstream users with essential and sensible guidelines on how to implement such a mechanism, which is similar and in the best case compatible with future upstream requirements.

Reply:
How about just trim it down to"

Accordingly, the core kernel MUST NOT assume any semantic meaning for completion codes or take actions based on their values besides printing error messages unless 
  - there is a specification of a particular application's completion code space written in a TRD, and 
  - the kernel can reliably identify that application and associate it with this specification.

Reply:
That works for me

Reply:
@phil-levis Just now catching up on the backlog. Sure, that sounds good and handles all of my concerns. Here's a suggestion which can be applied through GitHub:

```suggestion
Accordingly, the core kernel MUST NOT assume any semantic meaning for completion
codes or take actions based on their values besides printing error messages
unless

- there is a specification of a particular application's completion code space
  written in a TRD, and

- the kernel can reliably identify that application and associate it with this
  specification.
```

Reply:
I went ahead and applied the suggestion so that outstanding approvals will not get wiped when it is applied